core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -204,6 +211,14 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         // automatically, so try to tell the executor to stop itself. See SPARK-13519.
         executorDataMap.get(executorId).foreach(_.executorEndpoint.send(StopExecutor))
         removeExecutor(executorId, reason)
+        hostToExecutors.foreach( e => {
+          val execs = e._2
+          if (execs.contains(executorId)) execs -= executorId
+          if (execs.isEmpty) {
+            val host = e._1
+            hostToExecutors -= host
+          }
+        })
```

Comment:
if you switch to using `TaskSchedulerImpl.getExecutorsAliveOnHost` you won't need this at all, but some comments anyhow on this version just for the thought experiment:

1) this should be in `removeExecutor`, so you do it in all cases when the executor is removed
2) you can use a scala's PartialFunction syntax to unravel the pair a little more cleanly:

```scala
hostToExecutors.foreach { case (host, execs) =>
 ...
}
```